Ryotaro49/2023/eslint-rule-quotes-disable/index.md

```diff
@@ -0,0 +1,212 @@
+---
+title: 'ESLint で特定のファイルのルールを変更する方法'
+date:
+author: Ryotaro49
+tags: [ESLint, JavaScript]
+description: 'ESLint で特定のファイルのルールを変更する方法を紹介します。'
+---
+
+こんにちは、 Makky です。
+
+先日、とある JS ファイルを作った時に「**String must use singlequote**(文字列にはシングルクォートが必要ですよ！)」という Lint エラーがでてきていました。
+
+```js:title=Lintエラー
+ App •  ERROR  •  UI  Reported by ESLint:
+[eslint]
+/index.js
+    3:11  error  Strings must use singlequote  quotes
+```
+
+基本的に Lint エラーになっているのは、コーディング規約違反ですので、本来は修正すべきです。
+
+ただ、背景として自動生成したプログラムを組み込む際に、そのプログラムが Lint エラーを吐いていたので、 Lint エラーを無視させたいということがありました。
+
+そこで今回は、クォートを例として、特定のファイルのみ Lint エラーが出ないように設定する方法を紹介します。
+
+## 想定環境
+
+- エディター: VS code v1.79.2
+- Linter: ESLint v8.27.0
+
+## 設定方法
+
+以下のように、`overrides` を使用して新しいルールを記述します。
+
+ただし、すでに `rules` がある場合は、**`rules`の外に `overrides` を記述してください。**
+
+```js:title=ファイルパスを指定してルールを設定
+  overrides: [               // overrides を使用して新しくルールを作ります。
+    {
+      files: ["./src/index.js"], // ここでファイルを指定します。
+      rules: {
+        quotes: ["off"],
+      },
+    },
+  ],
+```
+
+`overrides` の中で `files` を記述し、ファイルを指定します。
+
+これで、ダブルクォートを使っても Lint エラーは出なくなりました。
+**指定したファイル**のみ `overrides` の中の `rules` のルールが適用されています。
+
+### プロジェクト全体とファイルごとで、異なるルールを適用したい場合
+
+`fruit.js` と`language.js` というファイルがあるとします。
+
+```js:title=fruit.js
+{
+  apple: 'りんご', // シングルクォート
+  grapes: "ぶどう", // ダブルクォート
+  banana: `バナナ`, // バッククォート
+}
+```
+
+```js:title=language.js
+{
+  JP: 'jp', // シングルクォート
+  US: "us", // ダブルクォート
+};
+```
+
+以下のコードは `rules` (プロジェクト全体) ではエラーが出るように、`overrides` (ファイルごと) では無効化するように設定しました。
+
+```js:title=eslintrc.js
+module.exports = {
+  root: true,
+
+  rules: {
+    quotes: ["error", "single"], // rules にシングルクォート以外を使うとエラーが出るように設定
+  },
+  overrides: [
+    {
+      files: ["fruit.js"],
+      rules: {
+        quotes: ["off"], // overrides にクォートのルールを無効化するように設定
+      },
+    },
+  ],
+};
+```
+
+```js:title=結果：Lintエラー
+ App •  ERROR  •  UI  Reported by ESLint:
+[eslint]
+/language.js
+  3:7  error  Strings must use singlequote  quotes  //"us", ダブルクォート
+```
+
+出力結果は、 *`overrides` で指定していないファイル*(language.js)でシングルクォート以外を使用しているため、エラーが出ます。
+
+ですが、**指定したファイル**(fruit.js)ではエラーが出ません。
+
+これで、*プロジェクト単位、ファイル単位*で**異なるルール**を設定できました。
+
+### プロパティの仕様
+
+以下は ESLint のエラーレベルと説明を表にまとめたものです。
+
+| エラーレベル | 説明                                                             |
+| ------------ | ---------------------------------------------------------------- |
+| `off`        | ルールを無効にする、エラーが表示されないためコードは実行される。 |
+| `warn`       | 警告として表示されるが、コードは実行される                       |
+| `error`      | エラーとして表示され、コードは実行されない                       |
```

Comment:
承知しました。ありがとうございます。
以下の表現にしようと思うのですが、いかがでしょうか。

- off : ルールを無効にする、終了コードは変更しない。（検査対象のプログラムは終了しない。）
- warn : 警告として表示されるが、終了コードは変更しない。（検査対象のプログラムは終了しない。）
- error : エラーとして表示され、終了コードを1（失敗）にする。（検査対象のプログラムは終了する。）

Reply:
「終了コードは変更しない」というよりは、「実行結果は成功として扱う」とするほうが良いかもしれません。
(すみません、このあたり、あまりいい日本語が思い浮かびませんでした...)

「検査対象のプログラムは終了(する|しない)。」は不要だと思います。
ややこしいですが、「検査対象のプログラムは」と書いてあると、 linter ではなくプログラム本体の実行の話になってしまいます。
あくまで linter ではソースコードにコーディング規約違反がないかをチェックしているのみで、プログラムの実行とは関係がありません。
終了コード、というのは、図の ③ の部分を指しています。

![linter-workflow](https://github.com/mseninc/blog/assets/18321329/943229c5-2213-410e-abaa-de1e71d05ce7)

補足ですが、終了コードの話は CI のテストやコミット前のチェック、 PR のマージの際に linter を走らせて、終了ステータスが 0 でない場合 (何らかのエラーがある場合) は、失敗として扱われること、警告は表示したいけれど、失敗として扱いたくない場合は `warn` を使えることが、ドキュメントに書いてあります。

> Rules are typically set to "error" to enforce compliance with the rule during continuous integration testing, pre-commit checks, and pull request merging because doing so causes ESLint to exit with a non-zero exit code.
>
> If you don’t want to enforce compliance with a rule but would still like ESLint to report the rule’s violations, set the severity to "warn".

Reply:
はい、この内容で問題ないと思います。

Reply:
ご確認ありがとうございます。
修正します。